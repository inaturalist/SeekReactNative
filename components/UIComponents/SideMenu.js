// @flow

import * as React from "react";
import { Image, Pressable, View } from "react-native";

import i18n from "../../i18n";
import { viewStyles, imageStyles, textStyles } from "../../styles/uiComponents/sideMenu";
import logoImages from "../../assets/logos";
import icons from "../../assets/icons";
import { capitalizeNames, setRoute } from "../../utility/helpers";
import { resetRouter } from "../../utility/navigationHelpers";
import { colors } from "../../styles/global";
import StyledText from "./StyledText";

type Props = {
  navigation: any
}

const SideMenu = ( { navigation }: Props ): React.Node => {
  const { navigate } = navigation;
  const menuItems = ["home", "achievements", "challenges", "observations", "inat", "about", "settings"];

  const navHome = ( ) => resetRouter( navigation );

  const navToPath = ( path ) => {
    if ( path === "Home" ) {
      navHome( );
    } else {
      if ( path === "Observations" ) {
        setRoute( "SideMenu" );
      }
      navigate( path );
    }
  };

  const renderMenuItems = ( ) => menuItems.map( ( item, i ) => {
    const titleCase = capitalizeNames( item ) || "";
    const name = i18n.t( `menu.${item}` ).toLocaleUpperCase();
    const path = item === "inat" ? "iNatStats" : titleCase;

    return (
      <Pressable
        key={item}
        accessibilityLabel={name}
        accessible
        onPress={() => navToPath( path )}
        style={[viewStyles.menuItem, i < menuItems.length - 1 && viewStyles.divider]}
      >
        {/* $FlowFixMe */}
        <Image
          source={icons[`menu${titleCase}`]}
          style={imageStyles.icon}
          tintColor={colors.menuItems}
        />
        <StyledText allowFontScaling={false} style={textStyles.text}>{name}</StyledText>
      </Pressable>
    );
  } );

  return (
    <View testID="side-menu" style={viewStyles.container}>
      <Pressable
        accessibilityLabel={i18n.t( "menu.home" )}
        accessible
        onPress={navHome}
      >
        <Image source={logoImages.seek} style={imageStyles.seekLogo} />
      </Pressable>
      {renderMenuItems( )}
    </View>
  );
};

export default SideMenu;
