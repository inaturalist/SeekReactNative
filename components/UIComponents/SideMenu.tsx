import * as React from "react";
import { Image, Pressable, View } from "react-native";

import i18n from "../../i18n";
import { viewStyles, imageStyles, textStyles } from "../../styles/uiComponents/sideMenu";
import logoImages from "../../assets/logos";
import icons from "../../assets/icons";
import { capitalizeNames, setRoute, StoredRoutes } from "../../utility/helpers";
import StyledText from "./StyledText";
import { baseTextStyles } from "../../styles/textStyles";

const SideMenu = ( { navigation } ) => {
  const { navigate, popTo } = navigation;
  const menuItems = ["home", "achievements", "challenges", "observations", "inat", "about", "settings"];

  const navToPath = ( path: string ) => {
    if ( path === "Observations" ) {
      setRoute( StoredRoutes.SideMenu );
    }
    navigate( path );
  };

  const renderMenuItems = ( ) => menuItems.map( ( item, i ) => {
    const titleCase = capitalizeNames( item ) || "";
    const name = ( item === "inat" ? "iNaturalist" : i18n.t( `menu.${item}` ) ).toLocaleUpperCase();
    const path = item === "inat" ? "iNatStats" : titleCase;

    return (
      <Pressable
        key={item}
        accessibilityLabel={name}
        accessible
        onPress={( ) => navToPath( path )}
        style={[viewStyles.menuItem, i < menuItems.length - 1 && viewStyles.divider]}
      >
        <Image
          source={icons[`menu${titleCase}`]}
          style={imageStyles.icon}
        />
        <StyledText allowFontScaling={false} style={[baseTextStyles.sideMenu, textStyles.text]}>{name}</StyledText>
      </Pressable>
    );
  } );

  return (
    <View testID="side-menu" style={viewStyles.container}>
      <Pressable
        accessibilityLabel={i18n.t( "menu.home" )}
        accessible
        // Home should definitively be in the stack, as it is the initial route of the app.
        // So we can just pop back to it.
        onPress={( ) => popTo( "Home" )}
      >
        <Image source={logoImages.seek} style={imageStyles.seekLogo} />
      </Pressable>
      {renderMenuItems( )}
    </View>
  );
};

export default SideMenu;
