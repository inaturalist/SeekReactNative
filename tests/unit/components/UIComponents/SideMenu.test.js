import React from "react";
import { render, screen } from "@testing-library/react-native";

import SideMenu from "../../../../components/UIComponents/SideMenu";

describe( "SideMenu", () => {
    test( "should render correctly", () => {
        render( <SideMenu navigation={{}} /> );
        screen.findByTestId( "side-menu" );
        expect( screen ).toMatchSnapshot( );
     } );
} );
