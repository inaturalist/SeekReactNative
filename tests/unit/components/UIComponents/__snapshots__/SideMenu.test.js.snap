// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideMenu should render correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#38976d",
      "flex": 1,
      "flexGrow": 1,
      "justifyContent": "center",
    }
  }
  testID="side-menu"
>
  <View
    accessibilityLabel="Home"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Image
      source={
        {
          "testUri": "../../../assets/logos/logo-seek-splash-2020.webp",
        }
      }
      style={
        {
          "alignSelf": "center",
          "height": 79,
          "marginVertical": 41,
          "resizeMode": "contain",
          "width": 223,
        }
      }
    />
  </View>
  <View
    accessibilityLabel="HOME"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "flex-start",
          "paddingVertical": 21,
        },
        {
          "borderBottomColor": "#63d4ab",
          "borderBottomWidth": 1,
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/icons/icon-menu-home.webp",
        }
      }
      style={
        {
          "height": 25,
          "marginHorizontal": 25,
          "resizeMode": "contain",
          "tintColor": "#63d4ab",
          "width": 27,
        }
      }
      tintColor="#63d4ab"
    />
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "#ffffff",
            "flex": 1,
            "fontFamily": "Whitney-Semibold",
            "fontSize": 18,
            "letterSpacing": 1,
            "paddingEnd": 4,
            "paddingTop": 3,
          },
        ]
      }
    >
      HOME
    </Text>
  </View>
  <View
    accessibilityLabel="ACHIEVEMENTS"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "flex-start",
          "paddingVertical": 21,
        },
        {
          "borderBottomColor": "#63d4ab",
          "borderBottomWidth": 1,
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/badges/observations/icon-badge-gold.webp",
        }
      }
      style={
        {
          "height": 25,
          "marginHorizontal": 25,
          "resizeMode": "contain",
          "tintColor": "#63d4ab",
          "width": 27,
        }
      }
      tintColor="#63d4ab"
    />
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "#ffffff",
            "flex": 1,
            "fontFamily": "Whitney-Semibold",
            "fontSize": 18,
            "letterSpacing": 1,
            "paddingEnd": 4,
            "paddingTop": 3,
          },
        ]
      }
    >
      ACHIEVEMENTS
    </Text>
  </View>
  <View
    accessibilityLabel="CHALLENGES"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "flex-start",
          "paddingVertical": 21,
        },
        {
          "borderBottomColor": "#63d4ab",
          "borderBottomWidth": 1,
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/icons/icon-menu-challenges.webp",
        }
      }
      style={
        {
          "height": 25,
          "marginHorizontal": 25,
          "resizeMode": "contain",
          "tintColor": "#63d4ab",
          "width": 27,
        }
      }
      tintColor="#63d4ab"
    />
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "#ffffff",
            "flex": 1,
            "fontFamily": "Whitney-Semibold",
            "fontSize": 18,
            "letterSpacing": 1,
            "paddingEnd": 4,
            "paddingTop": 3,
          },
        ]
      }
    >
      CHALLENGES
    </Text>
  </View>
  <View
    accessibilityLabel="MY OBSERVATIONS"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "flex-start",
          "paddingVertical": 21,
        },
        {
          "borderBottomColor": "#63d4ab",
          "borderBottomWidth": 1,
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/icons/icon-menu-observations.webp",
        }
      }
      style={
        {
          "height": 25,
          "marginHorizontal": 25,
          "resizeMode": "contain",
          "tintColor": "#63d4ab",
          "width": 27,
        }
      }
      tintColor="#63d4ab"
    />
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "#ffffff",
            "flex": 1,
            "fontFamily": "Whitney-Semibold",
            "fontSize": 18,
            "letterSpacing": 1,
            "paddingEnd": 4,
            "paddingTop": 3,
          },
        ]
      }
    >
      MY OBSERVATIONS
    </Text>
  </View>
  <View
    accessibilityLabel="INATURALIST"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "flex-start",
          "paddingVertical": 21,
        },
        {
          "borderBottomColor": "#63d4ab",
          "borderBottomWidth": 1,
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/logos/iNatStatsSpeciesDetail-Bird.webp",
        }
      }
      style={
        {
          "height": 25,
          "marginHorizontal": 25,
          "resizeMode": "contain",
          "tintColor": "#63d4ab",
          "width": 27,
        }
      }
      tintColor="#63d4ab"
    />
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "#ffffff",
            "flex": 1,
            "fontFamily": "Whitney-Semibold",
            "fontSize": 18,
            "letterSpacing": 1,
            "paddingEnd": 4,
            "paddingTop": 3,
          },
        ]
      }
    >
      INATURALIST
    </Text>
  </View>
  <View
    accessibilityLabel="ABOUT"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "flex-start",
          "paddingVertical": 21,
        },
        {
          "borderBottomColor": "#63d4ab",
          "borderBottomWidth": 1,
        },
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/icons/icon-menu-seek.webp",
        }
      }
      style={
        {
          "height": 25,
          "marginHorizontal": 25,
          "resizeMode": "contain",
          "tintColor": "#63d4ab",
          "width": 27,
        }
      }
      tintColor="#63d4ab"
    />
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "#ffffff",
            "flex": 1,
            "fontFamily": "Whitney-Semibold",
            "fontSize": 18,
            "letterSpacing": 1,
            "paddingEnd": 4,
            "paddingTop": 3,
          },
        ]
      }
    >
      ABOUT
    </Text>
  </View>
  <View
    accessibilityLabel="SETTINGS"
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "flex-start",
          "paddingVertical": 21,
        },
        false,
      ]
    }
  >
    <Image
      source={
        {
          "testUri": "../../../assets/icons/icon-menu-settings.webp",
        }
      }
      style={
        {
          "height": 25,
          "marginHorizontal": 25,
          "resizeMode": "contain",
          "tintColor": "#63d4ab",
          "width": 27,
        }
      }
      tintColor="#63d4ab"
    />
    <Text
      allowFontScaling={false}
      style={
        [
          {
            "textAlign": "left",
          },
          {
            "color": "#ffffff",
            "flex": 1,
            "fontFamily": "Whitney-Semibold",
            "fontSize": 18,
            "letterSpacing": 1,
            "paddingEnd": 4,
            "paddingTop": 3,
          },
        ]
      }
    >
      SETTINGS
    </Text>
  </View>
</View>
`;
